Full text search

vdimir opened the pull request:
Resolves #734

This PR suggests full text search implementation. 
There are two search backends [bleve](https://github.com/blevesearch/bleve) and ElasticSearch (via [go-elasticsearch](https://github.com/elastic/go-elasticsearch) client).

Also I have considered implementation of inverted index by hands. It is required only implementation `indexer` interface with one method `IndexDocument` because the maintenance (such as batching, startup/shutdown etc.) is already done in front of bleve.

I have performed kind of load testing. I downloaded a bunch of reddit comments (to be close to real data) converted it to remark backup format, restored it, and ran request from local computer to droplet with disabled ratelimiter. Script for downloading and processing data and python notebook with testing can be found in [here](https://gist.github.com/vdimir/662a76926717c49b8b554af79792211e).

I was testing on ~80.000 comments restored from backup generating and sending requests from 8 threads.

Elastic behaves good.

Bleve slow downs remark.
When search and comment requests were sent, comment had timing 400ms @ p99 and 100 ms @ p50, search – 1000ms @ p99 and 200 @ p50, total rps ~40.

Also when indexing was enabled and only commenting was performed, latency spikes slowed down requests to >1000 ms every 5-10 sec. But rest 95% requests are good (~100 ms), total rps ~100.
So, I think spikes are connected with batch flushing, but I think this place requires attention.

Actually I think Elastic is an overkill for remark and I have implemented it only to ensure that interfaces are extensible to support such things.

Also I feel that my implementation is weak at some points but I'm not sure which things need improvement. 
**So, all comments and review will be really appreciated**. Especially I worry about architecture and working with concurrency. I am ready to make any changes starting from minor fixes to changing architecture at all.

Also sorry for mistakes in text, ask me about everything, I will explain if something isn't clear.

paskal commented:
I have trouble reviewing this PR because of vendoring I suppose: my IDE [just refuses to show a list of files](https://youtrack.jetbrains.com/issue/IDEA-247243). Would you be so kind as to split your repo into a branch with vendoring and branch on top of it with just the code changes, create a PR with just the code changes and ping me for review there?

vdimir commented:
@paskal I removed vendor from this PR and pushed it to separate branch [full-text-search-vendor](https://github.com/vdimir/remark/tree/full-text-search-vendor). I haven't create PR for this separate branch yet, because maybe it is better to keep one PR and merge changes back after reviews? Or I'll create another PR later if we decide to do it.

paskal commented:
Sorry for not being clear, I wanted to have PR like one we have here to be a separate one on top of vendoring branch in **your** fork, so I would review it there. I'll try to look into it by tomorrow evening.

paskal commented:
I expect to have the ability to renew it this weekend.

paskal commented on backend/app/rest/api/rest_public.go:
```diff
@@ -447,6 +460,64 @@ func (s *public) loadPictureCtrl(w http.ResponseWriter, r *http.Request) {
+	query := r.URL.Query().Get("query")
+	sortBy := r.URL.Query().Get("sort")
+
+	limit := s.getNumericParam(r, "limit", 20)
```
For consistency, I would propose to go with the way other functions like `listCtrl` get numeric params.

paskal commented on backend/app/store/engine/bolt.go:
```diff
@@ -398,6 +398,15 @@ func (b *BoltDB) ListFlags(req FlagRequest) (res []interface{}, err error) {
 }
 
+// ListSites returns slice with all existing site ids
+func (b *BoltDB) ListSites() ([]string, error) {
```
Same as with example remote store, this should be covered by a test.

vdimir replied:
I've removed ListSites, because Sites already passed to search at creation stage

paskal commented on backend/app/store/service/service.go:
```diff
@@ -855,6 +874,61 @@ func (s *DataStore) Last(siteID string, limit int, since time.Time, user store.U
 	return s.alterComments(comments, user), nil
 }
 
+// Search commens using user query
```
`commens` typo

paskal commented on backend/app/store/service/service.go:
```diff
@@ -855,6 +874,61 @@ func (s *DataStore) Last(siteID string, limit int, since time.Time, user store.U
 }
 
+// Search commens using user query
+func (s *DataStore) Search(siteID, query, sortBy string, from, limit int) (*SearchResultPage, error) {
```
Part up to there (server.go -> rest_public.go -> service.go, engine changes) seems sane, I have no questions about the flow of data.
At this point I haven't reviewed only the new `search` module.

paskal replied:
UPD: except for `WrapEngine`.

paskal commented on backend/app/store/image/image_mock.go:
```diff
@@ -1,4 +1,4 @@
-// Code generated by mockery v1.0.0. DO NOT EDIT.
+// Code generated by mockery v1.1.1. DO NOT EDIT.
```
Doesn't seems relevant to this PR.

vdimir replied:
Remove this change

paskal commented on backend/app/cmd/server.go:
```diff
@@ -347,7 +358,15 @@ func (s *ServerCommand) newServerApp() (*serverApp, error) {
+	}
+
+	if searchService != nil {
+		storeEngine = search.WrapEngine(storeEngine, searchService)
```
This doesn't feel right, I'll write if I'll have any idea on how to make it better.
I think either storeEngine should already have search embedded 100% of the time with something like NoopSearchService if it's not present, or search should be a separate object with storeEngine passed to it explicitly and unaltered.

@umputun what do you think about this particular place?

paskal replied:
I think that with NoopSearchService it should be possible to do what you do in the wrapper in the normal Engine, even for a remote one. But then we'll the same logic in multiple places which is bad as well.

vdimir replied:
Should I wrap storeEngine with NoopSearchService ? How I should get wrappedEngine in such flow? Or I may try other approach without wrapping?

paskal replied:
I'll think about it. But wrapping shouldn't be conditional, there should be either real or Noop search engine and the flow should be the same for both.

vdimir replied:
I've create NoOp service but sore.Interface still wrapped. Proxy used in all cases whether search enabled or not (noop used)

paskal commented on backend/app/store/search/service.go:
```diff
@@ -0,0 +1,83 @@
+}
+
+// Help returns text doc for query language
+func Help(engineType string) string {
```
Exposing different kinds of search APIs to users as-is doesn't seem right, as then it would require frontend to support them all. And I don't know how safe it is: with the current design, I potentially can craft intentionally bad search terms and produce a denial of service attack.
My vote is to expose simpler API which would have limited capabilities but would be safe to use.
/cc @umputun

vdimir replied:
I'm not exposing kind of apis to frontend, this function is only used internally to get Usage info for current engineType and send it to frontend

vdimir replied:
Also I moved help method to search.Service interface

paskal commented on backend/app/store/search/search_test.go:
```diff
@@ -0,0 +1,333 @@
+	err := searcher.Init(context.Background(), nil)
+	assert.NoError(t, err)
+
+	searcher.IndexDocument("123456", &store.Comment{
```
`assert.NoError` is missing there

paskal commented on backend/app/store/search/search_test.go:
```diff
@@ -0,0 +1,333 @@
+		Timestamp: time.Date(2017, 12, 20, 15, 20, 28, 0, time.Local),
+	})
+
+	_ = searcher.Flush("test-site")
```
It doesn't seem right to not check for errors in such a simple to check place in test.

paskal commented on backend/app/store/search/search_test.go:
```diff
@@ -0,0 +1,333 @@
+	searcher, teardown := createTestService(t, sites)
+	defer teardown()
+
+	b, teardown := createDB(t, 42, sites)
```
Please rename teaddown functions to unique names as now you have two of them deferred and I am not sure if you won't call the same function (latter one) twice instead of calling two separate functions.

vdimir commented:
> Another round, I reviewed everything but `search` module as it's a beast on its own. @vdimir please let me know if what I wrote something which makes sense or I'm looking into wrong things and that's not what you were looking for.

Thanks for feedback!
I have read comments and understand how to fix most of them. I'll fix it or reply in conversation.

vdimir commented on backend/app/cmd/server.go:
```diff
@@ -513,6 +533,17 @@ func (a *serverApp) run(ctx context.Context) error {
 	go a.imageService.Cleanup(ctx) // pictures cleanup for staging images
 
+	if a.dataService.SearchService != nil {
+		go func() {
```
It may take a long time e.g. in case of cold startup with bunch of data in database

paskal replied:
Doing that with context should be enough then. But `all documents indexed` doesn't seem right in case there is an error or context is canceled.

vdimir commented on compose-dev-backend.yml:
```diff
@@ -37,6 +37,9 @@ services:
             - BACKUP_PATH=/srv/var/backup
+            - SEARCH_ENGINE
+            - SEARCH_ANALYZER
+            - SEARCH_INDEX_PATH=/srv/var/search
```
Added to all three docker-copose

vdimir commented on backend/app/store/search/elastic.go:
```diff
@@ -0,0 +1,361 @@
+	"path"
+	"strings"
+
+	"github.com/elastic/go-elasticsearch/v7"
```
It is sorted in blocks [stdlib] [github deps] as in other files. How it should be sorted?

paskal replied:
Imports from the same application shouldn't stay in the same block as third-party imports.

vdimir commented on backend/app/store/search/buffered_engine.go:
```diff
@@ -0,0 +1,271 @@
+				break
+			}
+		case *idxFlusher:
+			defer func() { val.notifier <- nil }()
```
I've put notifiers in array and touch then at the end of function in loop

vdimir commented on backend/app/store/search/elastic.go:
```diff
@@ -0,0 +1,361 @@
+			errs = multierror.Append(errs, errors.Wrapf(err, "error getting index status"))
+			continue
+		}
+		defer func() {
```
I thinks it is possible to close body immediately.

vdimir commented on backend/app/store/search/multiplexer.go:
```diff
@@ -0,0 +1,182 @@
+}
+
+// IndexDocument adds comment to index
+func (s *multiplexer) IndexDocument(commentID string, comment *store.Comment) error {
```
I've removed it from signature, comment.ID used everywhere instead

vdimir commented on backend/app/cmd/server.go:
```diff
@@ -957,6 +988,28 @@ func (s *ServerCommand) makeAuthenticator(ds *service.DataStore, avas avatar.Sto
+	}
+
+	if s.SearchEngine.IndexPath == "" {
+		return nil, errors.Errorf("wrong search index path %s", s.SearchEngine.IndexPath)
```
I've changed message to `search index path is not set`

vdimir commented on backend/app/rest/api/rest_public.go:
```diff
@@ -388,19 +391,29 @@ func (s *public) countMultiCtrl(w http.ResponseWriter, r *http.Request) {
 }
 
-// GET /list?site=siteID&limit=50&skip=10 - list posts with comments
+// GET /list?site=siteID&limit=20&skip=10 - list posts with comments
```
It is changed accidentally, all changes should be in searchQueryCtrl, I moved it to right place.

vdimir commented on backend/app/cmd/server.go:
```diff
@@ -347,7 +358,15 @@ func (s *ServerCommand) newServerApp() (*serverApp, error) {
+		return nil, errors.Wrap(err, "failed to make search service")
+	}
+
+	if searchService != nil {
```
If we wrap engine.Interface with Noop it adds one indect layer, do we actually need it? It can affect on performance also not so clear in my opinion

paskal replied:
Readability > performance. And I don't know how noop implementation could affect the performance.

vdimir commented on backend/app/rest/api/rest_public_test.go:
```diff
@@ -917,6 +920,99 @@ func TestRest_LastCommentsStreamSince(t *testing.T) {
+		defer cancel()
+		err = ds.SearchService.Init(ctx, ds.Engine)
+		require.NoError(t, err)
+		ds.Engine = search.WrapEngine(ds.Engine, ds.SearchService)
```
I've modified srv.DataStore

vdimir commented on backend/app/rest/api/rest_test.go:
```diff
@@ -333,17 +333,29 @@ func TestRest_cacheControl(t *testing.T) {
-	var testDB string
-	// pick a file name which is not in use for sure
+// randomPath pick a file or folder name which is not in use for sure
+func randomPath(tempDir, basename, suffix string) (string, error) {
```
It may be difficult to maintain two PRs, do we really need keep this changes separately?

paskal replied:
PR with these changes would be merged way quickier and you could rebase your changes here on top of master after that.
I'll be able to re-review this PR by tomorrow evening I suppose.

paskal replied:
My vote is still to move changes which are not related to full-text search to separate PR.

vdimir replied:
ref #765

paskal commented:
I should have time tonight to review it thoughtfully.

vdimir commented:
> I should have time tonight to review it thoughtfully.

Thanks!

I think main issue that I haven't figure out is wrapping store.Interface. Wrapping store with proxy is simple way to handle all store calls and perform additional work to support search without modifying store. But I understand that this solution not so clear, but I don't know how to do it better.

paskal commented on backend/_example/memory_store/accessor/data.go:
```diff
@@ -280,6 +280,15 @@ func (m *MemData) ListFlags(req engine.FlagRequest) (res []interface{}, err erro
 	return nil, errors.Errorf("flag %s not listable", req.Flag)
 }
 
+// ListSites returns slice with all existing site ids
```
I assume that had to be removed with others in ff41fd3d68becd54ec4482ac8a3eea687feaef10?

paskal reviewed (commented):
I'll look into search module and the architecture tomorrow. It would be great if you would fix small things and move unrelated changes out of PR, that would really make things simpler for me.

vdimir commented on backend/app/rest/api/rest_public_test.go:
```diff
@@ -925,3 +1021,13 @@ func postComment(t *testing.T, url string) {
+
+// equalsStringSliceUnordered chechs equality of two string sets
+// Note: passed slices are modified
+func equalsStringSliceUnordered(t *testing.T, expected, actual []string) {
```
Done, but it requires sorting slices to make comparison order-agnostic (like set equality)

vdimir commented on backend/app/cmd/server_test.go:
```diff
@@ -345,7 +348,18 @@ func TestServerApp_Failed(t *testing.T) {
 }
 
 func TestServerApp_Shutdown(t *testing.T) {
```
I can't find it, may you give some details?

paskal replied:
`go test -race -count 1 ./...` shows race conditions in `TestServerApp_Shutdown`, `TestSearch_OtherFields` and `TestSearch_SiteMux`.

vdimir replied:
Data race happens because I increment `shutdownWait` inside `indexDocumentWorker` and this write wasn't synchronized with `Wait` in `Close` method. So, I've move increment [outside](https://github.com/umputun/remark42/pull/757/files#diff-c78bc7caba2f8fa696a1de6869a8a875R92) of function, but caller (same module) should remember to call it, is it ok?


Also I wonder why there is no race with [this](https://github.com/umputun/remark42/pull/757/files#diff-9210fd29d3e63f5bdab3f203d850af54R89) access to shutdownWait. Is it true that it synchronizes via queueNotifier ?

paskal replied:
The link you provided is not valid, please quote the code. Race condition is gone as far as I can see.

paskal commented:
Very busy work week, I hope I'll have time for the review either on Friday or Sunday.

paskal commented on backend/app/store/search/buffered_engine.go:
```diff
@@ -0,0 +1,281 @@
+}
+
+// Init engine. It loads unindexed comments from ahead log saved from buffer on shutdown
+// Return true if engine initalizated before, false means cold start
```
typo, should be `initialized`

paskal commented on backend/app/store/search/buffered_engine.go:
```diff
@@ -0,0 +1,281 @@
+// Delete comment from index
+func (s *bufferedEngine) Delete(commentID string) error {
+	if err := s.index.Delete(commentID); err != nil {
+		return errors.Wrapf(err, "cannot detele comment %q from search index", commentID)
```
typo `detele`

paskal commented on backend/app/store/search/buffered_engine.go:
```diff
@@ -0,0 +1,281 @@
+}
+
+// bufferedEngine provides common functionality around searchEngine,
+// like buffering, startup/shotdown handling etc
```
typo `shotdown`

paskal commented on backend/app/store/search/bleve.go:
```diff
@@ -0,0 +1,214 @@
+		index, err = bleve.New(indexPath, createIndexMapping(analyzerMapping[analyzer]))
+	} else if errOpen == nil {
+		if !st.IsDir() {
+			return nil, errors.Errorf("index path shoule be a directory")
```
typo `shoule`

paskal commented on backend/app/store/search/elastic.go:
```diff
@@ -0,0 +1,359 @@
+	ready        bool
+}
+
+type elacticQuery struct {
```
I think there are a bunch of typos in this file, `elactic` -> `elastic`

paskal commented on backend/app/store/search/multiplexer.go:
```diff
@@ -0,0 +1,184 @@
+// Init creates missing indexes and index existing documents
+func (s *multiplexer) Init(ctx context.Context, e engine.Interface) error {
+	/* TODO(@vdimir)
+	 * This impmlementation could leave index inconsistent with storage in some rare cases.
```
`impmlementation` -> `implementation`

paskal reviewed (commented):
Generally looks good, aside from TODO stuff which I didn't check. Also, I didn't check test coverage as the code is not 100% complete at the moment, and I didn't run it yet as I want to run when the code review from Umputun will be passed.
I propose to fix review comments, ping me directly if you'll need any clarifications, and after these comments are resolved ping Umputun for final review. It's unlikely we'll jump on your todo items so I propose to add them as comment threads in review and ping me/Umputun directly on them if you want our reaction.

vdimir commented on backend/app/store/search/bleve.go:
```diff
@@ -0,0 +1,214 @@
+		indexPath:     indexPath,
+	}
+
+	go eng.indexDocumentWorker()
```
Comment is added

paskal commented on backend/app/store/search/buffered_engine.go:
```diff
@@ -0,0 +1,292 @@
+		}
+	}
+	if err != nil {
+		log.Printf("[ERROR] error %v writing log file", err)
```
We have multiple `err` reassigments inside the loop but only last one is printed, I think thig log.Printf should be incide the loop.

vdimir replied:
We assign it once in `err = dumpDoc(f, val)` and keep same because we skip processing of other elements after error
https://github.com/umputun/remark42/pull/757/files/f16eeba985b12c67591ff99fa96b7497359026c8#diff-9210fd29d3e63f5bdab3f203d850af54R156

vdimir commented on backend/app/store/search/buffered_engine.go:
```diff
@@ -0,0 +1,281 @@
+				continue
+			}
+			data = append(data, 0x0)
+			_, err = f.Write(data)
```
Actually we do not write data anymore if error was occurred, but we need to traverse all items in batch to send message to all waiters. I've perform slight refactoring of this part to make it more clear

vdimir commented on backend/app/store/search/bleve.go:
```diff
@@ -0,0 +1,214 @@
+	"russian":  bleveRu.AnalyzerName,
+}
+
+type bleveBatch struct {
```
I've split search module into `internal` and `service` (public part). Root module `search` depends on both and contains only decorator for engine and factory method that handle creation of service to `internal` part. In `internal` module structures from `service` are used and interfaces from it are implemented.

vdimir commented on backend/app/store/search/multiplexer.go:
```diff
@@ -0,0 +1,184 @@
+}
+
+// Flush documents buffer for site
+// If `Flush` called before `Init` or after `Init` that ends with error it blocks forever
```
Yes, I decide to return error if Flush called before ready, so caller need to handle it.

vdimir commented on backend/app/store/search/internal/elastic.go:
```diff
@@ -0,0 +1,364 @@
+
+// Flush all unprocessed documents
+func (e *elastic) Flush(siteID string) error {
+	// TODO(@vdimir)
```
Can we leave this part for now? Actually I don't think that elastic would be used in production, but implementing it not so clear. So, can supporting flush for elastic be omitted for now?

paskal replied:
Yes, but please explicitly state what code there should be and why it's not written so it would be relatively easy to discover and fix if somebody will need that.

vdimir replied:
Ok, I've added comment and return error from this method

vdimir commented on backend/app/store/search/internal/multiplexer.go:
```diff
@@ -0,0 +1,190 @@
+
+// Init creates missing indexes and index existing documents
+func (s *multiplexer) Init(ctx context.Context, e engine.Interface) error {
+	/* TODO(@vdimir)
```
Actually now we call `Init` in synchronous way (https://github.com/umputun/remark42/pull/757#discussion_r467503439), so new document won't be handled during `Init`

paskal replied:
So this is not actual anymore?

paskal replied:
To clarify what I mean: this function should be removed from the current code as far as I can see, that's the only action item in that thread.

vdimir replied:
It's required to initialize searcher at startup, read non-indexer documents, and important for first start. How it can be removed?

paskal replied:
Sorry, I misunderstood you and the code. Nothing to do here then.

vdimir commented on backend/app/store/search/internal/buffered_engine.go:
```diff
@@ -0,0 +1,296 @@
+// Init engine. It loads dumped comments from ahead log saved from buffer on shutdown
+// Return true if engine initialized before, false means cold start
+func (s *bufferedEngine) Init(ctx context.Context) (bool, error) {
+	// TODO(@vdimir) add tests for this part
```
I feel that this part need to be tested, but not sure how. I'll try to do something

paskal replied:
Generally, it's a good idea to create real instance of something and test internal behavior using only public methods. That way it's easy to refactor if needed and easy to replicate in case you'll add new search engine.

vdimir replied:
This part is difficult to test as black box. And for this I've written special test

vdimir reviewed (commented):
I added issue to my TODOS

paskal commented on backend/app/store/search/internal/bleve.go:
```diff
@@ -0,0 +1,235 @@
+	return eng, nil
+}
+
+// NewBleveService create search sservice based on bleve engine
```
```suggestion
// NewBleveService create search service based on bleve engine
```

paskal commented on backend/app/store/engine/engine_mock.go:
```diff
@@ -1,4 +1,4 @@
-// Code generated by mockery v1.0.0. DO NOT EDIT.
+// Code generated by mockery v1.1.1. DO NOT EDIT.
```
```suggestion
// Code generated by mockery v1.0.0. DO NOT EDIT.
```

paskal commented on backend/app/store/search/service/search_service.go:
```diff
@@ -0,0 +1,89 @@
+package search
```
This should be in the root folder, otherwise project won't compile:
```
app/store/service/service.go:40:25: undefined: "github.com/umputun/remark42/backend/app/store/search".Service
```

paskal commented on backend/app/store/search/internal/bleve.go:
```diff
@@ -0,0 +1,235 @@
+}
+
+// NewBleveService create search sservice based on bleve engine
+func NewBleveService(params service.SearcherParams) (s service.Service, err error) {
```
import cycle. You should return `Bleve` object which signature satisfies `service.Service`, and have no mentions of `service.*` in internal module to fix that.

paskal commented on backend/app/store/search/internal/multiplexer.go:
```diff
@@ -0,0 +1,190 @@
+	return s.ready.Load().(bool)
+}
+
+func indexTopic(ctx context.Context, comments []store.Comment, e engine.Interface, s indexer) (int, *multierror.Error) {
```
`e engine.Interface` is unused here.

paskal commented on backend/app/store/search/internal/multiplexer.go:
```diff
@@ -0,0 +1,190 @@
+		cnt, topicErrs := indexTopic(ctx, comments, e, s)
+		indexedCnt += cnt
+
+		if errs = multierror.Append(errs, topicErrs.Errors...); errs.Len() >= maxErrsDuringStartup {
```
My editor told me topicErrs might be nil, which will result in an error.

paskal commented on backend/app/cmd/server_test.go:
```diff
@@ -586,6 +600,84 @@ func TestServer_loadEmailTemplate(t *testing.T) {
+
+	go func() { _ = app.run(ctx) }()
+	waitForHTTPServerStart(port)
+	app.dataService.SearchService.Flush("remark")
```
```suggestion
	assert.NoError(t,app.dataService.SearchService.Flush("remark"))
```

paskal commented on backend/app/store/search/internal/buffered_engine.go:
```diff
@@ -0,0 +1,296 @@
+	return path.Join(s.indexPath, aheadLogFname)
+}
+
+// dumpDoc writes document to file separeted with \0
```
```suggestion
// dumpDoc writes a document to file separated with \0
```

paskal commented on backend/app/store/search/internal/buffered_engine.go:
```diff
@@ -0,0 +1,296 @@
+			return errors.Errorf("reading ahead log interrupted")
+		default:
+		}
+		// read documents separeted with \0
```
```suggestion
		// read documents separated with \0
```

paskal commented on backend/app/store/search/internal/multiplexer.go:
```diff
@@ -0,0 +1,190 @@
+// Flush documents buffer for site
+func (s *multiplexer) Flush(siteID string) error {
+	if !s.Ready() {
+		return errors.New("not initalized")
```
```suggestion
		return errors.New("not initialized")
```

paskal commented on backend/app/rest/api/rest_public.go:
```diff
@@ -447,6 +451,80 @@ func (s *public) loadPictureCtrl(w http.ResponseWriter, r *http.Request) {
 }
 
+// GET /search?site=siteID&query=queryText&limit=20&skip=10 - search documents
+func (s *public) searchQueryCtrl(w http.ResponseWriter, r *http.Request) {
```
This tested only partly, I think it's worth effort improving the coverage.

vdimir replied:
I've improve coverage

paskal commented on backend/app/store/search/internal/bleve.go:
```diff
@@ -0,0 +1,235 @@
+		if err != nil {
+			return nil, errors.Wrap(err, "cannot open index")
+		}
+	case errOpen == nil:
```
Testing this and default case as well would be good.

paskal commented on backend/app/store/search/internal/elastic.go:
```diff
@@ -0,0 +1,364 @@
+// elastic implements Service directly (not searchEngine)
+// because it has own mechanisms for batching
+// and do not require additional maintenance
+type elastic struct {
```
This file is not tested at all, it shouldn't be like that.

vdimir replied:
I've added some simple straight test, now coverage is about 40%. I've mocked elastic, but haven't mock meaningful response, it always returns 200 with empty body. But we test that requests are sent.

paskal commented on backend/app/store/search/store_engine_decorator.go:
```diff
@@ -0,0 +1,58 @@
+)
+
+// StoreEngineDecorator proxies requests to store/engine.Interface and index incoming data
+type StoreEngineDecorator struct {
```
This file content is very important part of the code and they should be tested as well.

vdimir replied:
I've added mock for searcher and test with it

paskal reviewed (commented):
Excellent job! Few very minor things to fix plus few tests to write, and also please check that it compiles as it doesn't now. After fixing that review notes ping Umputun directly, I don't think I have any more things I'm conserned about other than ones written in that review comments.

vdimir commented:
> Excellent job! Few very minor things to fix plus few tests to write, and also please check that it compiles as it doesn't now. After fixing that review notes ping Umputun directly, I don't think I have any more things I'm conserned about other than ones written in that review comments.

Thanks!
A lot of comments connected with typos, and it is not the thing that is required checking from reviewer, sorry for that, I'll more careful and will try to setup spellchecker or something like that to not disturb reviewer with such minors.

I hope I am finishing fixing latest issues soon.

vdimir commented:
I finally run service after all modification, load bunch of comments like in first post in this PR, seems it works.
Should I merge vendor files? Now it is in separate branch [full-text-search-vendor](https://github.com/vdimir/remark/tree/full-text-search-vendor)

vdimir commented:
@umputun could you review the code, please?

umputun commented:
Sorry for the long delay and thx for the impessive work and detailed reviews.

I have tried to understand the general structure/flow before I dig deeper into implementation details and it got me confused a little bit on multiple levels. Somehow the overal structure doesn't look intutive to me and I found myself looking at things implementing unexpected logic in unexpected places. Part of my confusion represented in the comments I've made, but I'm not really sure if my comments make much sense due to the general confusion.

Fundametnally all of this feels to me as two levels of abstraction, Servcie and Engine:

- search.Service - this is a struct (interfaces could be defined by consumers) providing all high-level API for the search functionality the end consumer (i.e. REST) needs. I'd expect some kind of Search call here. Probably nothing else needed, at least logically.

- both engines (bleve and elastic) are some implementations of search.Engine interface defined near search.Service

- the logic currently done by multiplexer, i.e. different sites support, probably should be the core part of the search.Service

- the service should be used directly by REST at least to perform search. It shoudln't be embeded into  servcie.DataStore

- the other part of search.Service (extracted interface) reponsible for create/updates/delete can be integrated into service.DataStore so, the current Create/Edit/Delete calls will affect search indexes properly

- The otheral package structure could represend this as a couple of sub-packages - engine and service or search_engine and search_service to make imports clear and avoid conflicts. On the top level (i.e search.go) we can keep all shared things used by both packages and this way import loops could be prevented.

and just to reiterate the general flow from the consumer point of view: the only thing consumers care about and use are interfaces extracted from the service struct. Some consumers may need different subset of those methods.

vdimir commented:
@umputun , thank you for review!

I think I got general idea and I'm going to rewrite PR reusing my code with respect to your notes. I'll ask if some additional question occurs during implementation. I hope I'll to start soon :)

paskal commented:
@vdimir I would love to have your work incorporated into the next release (1.7.0), please let us know if there is any way to assist you.

vdimir commented:
> @vdimir I would love to have your work incorporated into the next release (1.7.0), please let us know if there is any way to assist you.

Thanks for the concern. I'm going to work on it this weekend, I think after that we can discuss it more detailed.
Also when new release is planned to happen?

paskal commented:
After these changes + documentation ready, + some UI changes if we'll find someone to make them in time.

vdimir commented:
Sorry, I've started refactoring, but haven't enought time to work on it. I really want to finish, but I'll postpone it to the end of month or beginnig of next month.

paskal commented:
@vdimir, feel free to ping me in the email or telegram if you need help with this one; I'm fairly interested in getting this merged.

vdimir commented:
I tried to reorganize code according to this comment https://github.com/umputun/remark42/pull/757#issuecomment-720160517 

Please, look at the code to check if current understating of `search.Service` and `search.Engine` correspond to your sense.

Currently I removed all stuff connected with batching and also it doesn't perform indexing of existing documents on stratup, I'll add it later if current structure is ok.

Also, I removed elastic engine, not sure that it should be supported.

cc @umputun @paskal

umputun commented:
looks like you missed some vendor updates

paskal commented:
The vendor update is not here because I asked that initially. Otherwise, it's harder to review.

paskal commented on backend/app/rest/api/rest.go:
```diff
@@ -354,13 +358,17 @@ func (s *Rest) controllerGroups() (public, private, admin, rss) {
 	}
-
+	if s.SearchService != nil {
+		// we can't assign without checking for nil because we can get not nil interface containing nil underline
```
Sounds strange, would be nice to avoid this check if possible.

vdimir replied:
It's about this feature, when we assign null pointer to struct to interface

https://play.golang.org/p/OS8a9wbG0dJ

paskal reviewed (commented):
I do not see any red flags. The current composition looks logical to me.

vdimir commented on backend/app/store/search/bleve_engine.go:
```diff
@@ -0,0 +1,207 @@
+	batch := b.index.NewBatch()
+
+	for _, comment := range comments {
+		err := batch.Index(comment.ID, comment)
```
Not sure that we should index `store.Comment` directly, maybe it's better to convert it to simpler plain structure and store it to index.

It will simplify bleve mapping for struct (see `createIndexMapping`), and we won't need to specify `user.name` (instead of `user` or `username` if we put it on top level of struct) when we query for username.

Like it was in first version
https://github.com/umputun/remark42/blob/d755ebaca2361216f06813d8f8b6d5d953f4bdc4/backend/app/store/search/types/types.go#L72-L82

vdimir commented:
From my side only plans to add buffering of new documents before indexing is left. And I'm not sure that it's required now.

paskal commented:
@vdimir, I propose to get it to the MVP state with as limited functionality as you think would be enough to start, then we'll merge it, and it will be possible to improve it and to build a frontend part for it, which is way more challenging when it's not in master.

Could you please finalize the PR to that state and create follow-up issues for frontend support of the search and the parts you didn't complete here so that they won't be lost but will be documented?

vdimir commented:
> LGTM with few minor things here and there, tests part for search.go could be ommited if too difficult as it's tested from another location.

@paskal thanks for taking a look, all comments are resolved

paskal reviewed (dismissed):
LGTM. Examples should be added to `backend/remark.rest` as well, but that could be done in a separate PR: no need to block this one anymore.

vdimir commented:
@paskal @umputun 
I've resumed work on the feature and pushed the branch with the updated code. I tried to simplify it and added more comments about the high-level API (please point me to the places that are still not documented). May you please take a look?

vdimir commented:
> build / Build Docker images (pull_request) Successful in 3m 

Should it be more checks?

umputun commented on backend/app/cmd/server.go:
```diff
@@ -637,6 +652,11 @@ func (a *serverApp) run(ctx context.Context) error {
 		log.Printf("[WARN] failed to resubmit comments with staging images, %s", e)
 	}
 
+	for _, siteID := range a.Sites {
```
this unbounded concurrency makes me worry a little bit. In some extreme cases user may have a lot of sites

I suggest using SizedGroup here and in other places if you need to limit it with error handling - ErrSizedGroup from https://github.com/go-pkgz/syncs 

this package already in dependency and used by other parts of code to do a similar thing

vdimir replied:
I've changed it to use `ErrSizedGroup`. It's global for all sites, and one worker processes one topic.

umputun commented on backend/app/rest/api/rest_public.go:
```diff
@@ -386,6 +387,67 @@ func (s *public) telegramQrCtrl(w http.ResponseWriter, r *http.Request) {
+// sort - sort by specified field, can be prefixed with "-" to reverse the order
+func (s *public) searchQueryCtrl(w http.ResponseWriter, r *http.Request) {
+	validateAndGetParam := func(value *int, from, to, defaultVal int, name string) bool {
+		if v, err := strconv.Atoi(r.URL.Query().Get(name)); err == nil {
```
generally I preffer to avoid else branches if possible. Her setting defaultVal prior to the if will achieve the same result but without else thing

umputun commented on backend/app/rest/api/rest_public.go:
```diff
@@ -386,6 +387,67 @@ func (s *public) telegramQrCtrl(w http.ResponseWriter, r *http.Request) {
+// skip - number of documents to skip
+// sort - sort by specified field, can be prefixed with "-" to reverse the order
+func (s *public) searchQueryCtrl(w http.ResponseWriter, r *http.Request) {
+	validateAndGetParam := func(value *int, from, to, defaultVal int, name string) bool {
```
maybe instead of modifying value and passing the ref, just return it?

umputun commented on backend/app/rest/api/rest_public.go:
```diff
@@ -386,6 +387,67 @@ func (s *public) telegramQrCtrl(w http.ResponseWriter, r *http.Request) {
+		if from <= *value && *value <= to {
+			return true
+		}
+		rest.SendErrorJSON(w, r,
```
this is unexpected for a nested function. I would rather return error and handle it by caller

umputun commented on backend/app/rest/api/rest_test.go:
```diff
@@ -207,6 +209,61 @@ func TestRest_rejectAnonUser(t *testing.T) {
+			err = json.Unmarshal(body, &result)
+			require.NoError(t, err)
+
+			require.Len(t, result.Comments, tt.count)
```
the test is not checking if the correct comment found, but only how many comments

vdimir replied:
Updated to check IDs as well

umputun commented on backend/app/rest/api/rest_test.go:
```diff
@@ -207,6 +209,61 @@ func TestRest_rejectAnonUser(t *testing.T) {
+			require.Len(t, result.Comments, tt.count)
+		})
+	}
+}
```
there is no test using offset/limit, and there are no test for unhappy path. Would be good to have them all added

vdimir replied:
How sophisticated should it be? In this test, we need to test that it works trough the query params, so a test with three comments is suitable, is it?

umputun commented on backend/app/store/service/service.go:
```diff
@@ -104,6 +107,10 @@ func (s *DataStore) Create(comment store.Comment) (commentID string, err error)
 	s.submitImages(comment)
 
+	if e := s.indexComment(comment); e != nil {
+		log.Printf("[ERROR] failed to index comment, %v", e)
```
I don't think you meant to error here and terminate the flow. This failure is not critical, should be WARN

umputun commented on backend/app/store/service/service.go:
```diff
@@ -516,6 +523,11 @@ func (s *DataStore) EditComment(locator store.Locator, commentID string, req Edi
 
 	err = s.Engine.Update(comment)
+
+	if e := s.indexComment(comment); e != nil {
```
same as above - should be WARN

umputun commented on backend/app/store/service/service.go:
```diff
@@ -890,6 +902,44 @@ func (s *DataStore) Last(siteID string, limit int, since time.Time, user store.U
+	comments := make([]store.Comment, len(seqrchRes.Keys))
+	for i, docKey := range seqrchRes.Keys {
+		// ignore error because comment can be deleted from the store
+		comments[i], _ = s.Get(docKey.Locator, docKey.ID, nonAdminUser)
```
this one is unclear. if Get doesn't return a valid comment, why we even put it to result? Maybe should be conditional append ?

vdimir replied:
We do not delete anything from the search index but expect to have an error for the deleted comments. But this code was a bit wrong, we need to skip it, and now it's done.

umputun commented on backend/app/store/service/service.go:
```diff
@@ -996,3 +1049,13 @@ func (s *DataStore) getSecret(siteID string) (secret string, err error) {
+	if s.SearchService != nil {
+		err := s.SearchService.Index(comment)
+		if err != nil {
+			return err
```
I'm not sure if it makes sense (have not checked what the error is), but maybe we want to wrap the error to indicate what comment failed to index?

umputun commented on backend/app/main_test.go:
```diff
@@ -158,5 +158,6 @@ func TestMain(m *testing.M) {
 		m,
 		goleak.IgnoreTopFunction("github.com/umputun/remark42/backend/app.init.0.func1"),
 		goleak.IgnoreTopFunction("net/http.(*Server).Shutdown"),
+		goleak.IgnoreTopFunction("github.com/blevesearch/bleve_index_api.AnalysisWorker"),
```
probably there is a reason for this ignore, but it will be great to add a comment explaining why it ign ored

vdimir replied:
Added comment. Actually, we can also shutdown this worker by calling `belve.Config.Shutdown()`, but it's a global worker, and it would be stopped forever. So, I suppose we can leave it as is, and killing these goroutines in `main` doesn't makes much sense.

umputun reviewed (changes requested):
Hi. Thx for the hard work, it looks impressive. I did a quick review of the caller/service side so far (have not checked the search package yet)

vdimir commented on backend/app/cmd/server.go:
```diff
@@ -264,6 +266,13 @@ type RPCGroup struct {
+type SearchGroup struct {
+	Engine    bool   `long:"engine" env:"ENGINE" description:"enable search engine"`
+	IndexPath string `long:"index_path" env:"INDEX_PATH" description:"search index location" default:"./var/search_index"`
+	Analyzer  string `long:"analyzer" env:"ANALYZER" description:"text analyzer type, set language-specific one to improve search quality" default:"standard"`
```
The list is pretty long (choice for each supported language, it's nine for now), but I've added it.

vdimir commented:
Sometimes, there's a check with the test, but it disappears, and only `build / Build Docker images (pull_request)` is shown, so not sure about the current state.

It should be two checks:

build

https://github.com/umputun/remark42/blob/363e05d58065939b647ad199deeab44f677a7213/.github/workflows/ci-build.yml#L17-L20

and test

https://github.com/umputun/remark42/blob/363e05d58065939b647ad199deeab44f677a7213/.github/workflows/ci-backend.yml#L12-L16

paskal commented on backend/app/store/service/service.go:
```diff
@@ -890,6 +897,49 @@ func (s *DataStore) Last(siteID string, limit int, since time.Time, user store.U
+
+	comments := make([]store.Comment, 0, len(seqrchRes.Keys))
+	for _, docKey := range seqrchRes.Keys {
+		comment, err := s.Get(docKey.Locator, docKey.ID, nonAdminUser)
```
`err` is shadowed here, would be nice to change it to `e` just in case.

paskal commented on backend/app/store/service/service.go:
```diff
@@ -890,6 +897,49 @@ func (s *DataStore) Last(siteID string, limit int, since time.Time, user store.U
+		SortBy: sortBy,
+	}
+
+	seqrchRes, err := s.SearchService.Search(req)
```
Little typo: `seqrchRes` -> `searchRes`

paskal commented on backend/app/store/service/service.go:
```diff
@@ -996,3 +1049,13 @@ func (s *DataStore) getSecret(siteID string) (secret string, err error) {
 	return secret, nil
 }
+
+func (s *DataStore) indexComment(comment store.Comment) error {
```
I think naming it `indexCommentForSearch` might be better for code readability.

paskal replied:
The reason is that it's outside the search module, and you need to look into the code of the function to know it's calling the search module.

paskal commented on backend/app/store/search/bleve.go:
```diff
@@ -0,0 +1,242 @@
+		bReq.SortBy([]string{"-time"})
+	case req.SortBy == "time" || req.SortBy == "-time" || req.SortBy == "+time":
+		bReq.SortBy([]string{req.SortBy})
+	default:
```
Please add test covering that case as well.

vdimir replied:
Added to search_test https://github.com/umputun/remark42/pull/757/commits/3e4515b0cbecc0927cfc3a47defbee1f3dd3675e#diff-f933ab083e2ed57c319100cfb2bc25f3fc8493387c6b2fd5531600d3150870aaR109

And also there's case in rest tests https://github.com/umputun/remark42/pull/757/commits/3e4515b0cbecc0927cfc3a47defbee1f3dd3675e#diff-f09c2502b9be06e17728dceb28af8a6a4c70e1db2cadfa1fd62b9d91af7f1897R253

paskal commented on backend/app/store/search/bleve.go:
```diff
@@ -0,0 +1,242 @@
+
+		bleveAnalyzerName, has := analyzerNameMapping[analyzer]
+		if !has {
+			return nil, fmt.Errorf("unknown analyzer %s", analyzer)
```
Please cover this case with test.

vdimir replied:
Added in separate test in search module https://github.com/umputun/remark42/pull/757/commits/3e4515b0cbecc0927cfc3a47defbee1f3dd3675e#diff-f933ab083e2ed57c319100cfb2bc25f3fc8493387c6b2fd5531600d3150870aaR261

paskal commented on backend/app/store/search/search_test.go:
```diff
@@ -0,0 +1,281 @@
+	return createTestServiceInDir(t, sites, idxPath, false)
+}
+
+func TestSearch_Site(t *testing.T) {
```
`skip` seems to be missing

vdimir replied:
Added https://github.com/umputun/remark42/pull/757/commits/3e4515b0cbecc0927cfc3a47defbee1f3dd3675e#diff-f933ab083e2ed57c319100cfb2bc25f3fc8493387c6b2fd5531600d3150870aaR96

paskal reviewed (approved):
Only nitpick comments, also requires rebase on top of current master.
Overall LGTM, I like the interfaces and way existing code interacts with the search module. @umputun please take a look.

paskal commented on backend/app/store/service/service.go:
```diff
@@ -154,12 +162,6 @@ func (s *DataStore) Get(locator store.Locator, commentID string, user store.User
 	return s.alterComment(c, user), nil
 }
 
-// Put updates comment, mutable parts only
```
Please, return it. For context, please see #1528.

vdimir replied:
https://github.com/umputun/remark42/pull/757/commits/3e4515b0cbecc0927cfc3a47defbee1f3dd3675e#diff-be11453fbc15fb371635d660cf48b74ef5c375e7ad0368e26a94ac4ba431e67bR165

vdimir commented on backend/app/rest/api/rest_public.go:
```diff
@@ -386,6 +387,73 @@ func (s *public) telegramQrCtrl(w http.ResponseWriter, r *http.Request) {
+	var err error
+
+	var limit int
+	if limit, err = getIntParamInRange(1, 100, 20, "limit"); err != nil {
```
Added to comment https://github.com/umputun/remark42/pull/757/commits/3e4515b0cbecc0927cfc3a47defbee1f3dd3675e#diff-100021e9c1bd23b43c8c4b504d3fa7e4c4df8b403f6c422fe22c6c038bf5ca50R393-R394
and to api doc https://github.com/umputun/remark42/pull/757/commits/3e4515b0cbecc0927cfc3a47defbee1f3dd3675e#diff-401d8deb237f494d4c60e7331c6cb5d1b3c7a3886f752a81c9a04d3df3c1b6ecR246

vdimir commented on site/src/docs/contributing/api/index.md:
```diff
@@ -233,3 +233,19 @@ http://oldsite.com/from-old-page/1 https://newsite.com/to-new-page/1
+
+* `GET /search?site=siteID&query=query` - search comments for query `query` and returns `ResultPage`
+* `GET /search?site=siteID&query=query&sort=-time` - sort by time ascending
+* `GET /search?site=siteID&query=query&sort=-time&limit=20&skip=10` - pagination, get 20 comments, skip 10 first
```
Done, and actually this example for descending, but now there're both with clarification which is _oldest_ and _newest_

vdimir commented on site/src/docs/configuration/parameters/index.md:
```diff
@@ -128,6 +128,9 @@ services:
 | ssl.key                        | SSL_KEY                        |                          | path to the key.pem file                                  |
 | ssl.acme-location              | SSL_ACME_LOCATION              | `./var/acme`             | dir where obtained le-certs will be stored                |
 | ssl.acme-email                 | SSL_ACME_EMAIL                 |                          | admin email for receiving notifications from LE           |
+| search.engine                  | SEARCH_ENGINE                  | `false`                  | enable search engine                                      |
```
Makes sense. After change noticed that it is not tested. Should we tests this configuration settings?

vdimir commented:
@paskal I updated the branch removing the accidental downgrading of two modules.

vdimir commented:
@paskal I updated the code to have synchronous indexing for a cold start in `(serverApp).run`. Internally it's still executed with several workers that process topics in parallel in a sized group. Performance in sufficient to block once on update here. 

Tested on 8cpu/16gb digitalocean droplet, dataset with 70k comments, took 12 sec:

<details>
<summary>Logs</summary>

```
remark42-dev  | 2023/04/02 14:30:20.571 [INFO]  {cmd/server.go:1224 cmd.(*ServerCommand).makeSearchService} creating search service
remark42-dev  | 2023/04/02 14:30:20 [INFO] creating new search index var/search_index/72656d61726b811c9dc5
remark42-dev  | 2023/04/02 14:30:20.587 [WARN]  {cmd/server.go:654 cmd.(*serverApp).run} admin basic auth enabled
remark42-dev  | 2023/04/02 14:30:20.587 [INFO]  {service/service.go:962 service.(*DataStore).IndexSites} start building search index for 1 sites
remark42-dev  | 2023/04/02 14:30:20.587 [INFO]  {migrator/backup.go:26 migrator.AutoBackup.Do} activate auto-backup for remark under ./var/backup, duration 24h0m0s
remark42-dev  | 2023/04/02 14:30:20.587 [DEBUG] {migrator/backup.go:29 migrator.AutoBackup.Do} first backup for remark at 2023-04-03 14:30:20.587640815 -0500 CDT m=+86400.821362338
remark42-dev  | 2023/04/02 14:30:20.588 [INFO]  {provider/dev_provider.go:49 provider.(*DevAuthServer).Run} run local oauth2 dev server on 8084, redirect url=http://127.0.0.1:8080/auth/dev/callback
remark42-dev  | 2023/04/02 14:30:23 [INFO] 3669 documents indexed from topic {remark https://remark42.com/demo/}
remark42-dev  | 2023/04/02 14:30:23 [INFO] 3667 documents indexed from topic {remark https://remark42.com/demo9/}
remark42-dev  | 2023/04/02 14:30:23 [INFO] 3652 documents indexed from topic {remark https://remark42.com/demo1/}
remark42-dev  | 2023/04/02 14:30:23 [INFO] 3736 documents indexed from topic {remark https://remark42.com/demo14/}
remark42-dev  | 2023/04/02 14:30:23 [INFO] 3648 documents indexed from topic {remark https://remark42.com/demo11/}
remark42-dev  | 2023/04/02 14:30:24 [INFO] 3720 documents indexed from topic {remark https://remark42.com/demo17/}
remark42-dev  | 2023/04/02 14:30:24 [INFO] 3702 documents indexed from topic {remark https://remark42.com/demo10/}
remark42-dev  | 2023/04/02 14:30:24 [INFO] 3729 documents indexed from topic {remark https://remark42.com/demo16/}
remark42-dev  | 2023/04/02 14:30:26 [INFO] 3653 documents indexed from topic {remark https://remark42.com/demo19/}
remark42-dev  | 2023/04/02 14:30:26 [INFO] 3702 documents indexed from topic {remark https://remark42.com/demo18/}
remark42-dev  | 2023/04/02 14:30:26 [INFO] 3750 documents indexed from topic {remark https://remark42.com/demo15/}
remark42-dev  | 2023/04/02 14:30:26 [INFO] 3724 documents indexed from topic {remark https://remark42.com/demo3/}
remark42-dev  | 2023/04/02 14:30:29 [INFO] 3686 documents indexed from topic {remark https://remark42.com/demo2/}
remark42-dev  | 2023/04/02 14:30:29 [INFO] 3782 documents indexed from topic {remark https://remark42.com/demo20/}
remark42-dev  | 2023/04/02 14:30:29 [INFO] 3774 documents indexed from topic {remark https://remark42.com/demo12/}
remark42-dev  | 2023/04/02 14:30:33 [INFO] 3623 documents indexed from topic {remark https://remark42.com/demo6/}
remark42-dev  | 2023/04/02 14:30:33 [INFO] 3688 documents indexed from topic {remark https://remark42.com/demo8/}
remark42-dev  | 2023/04/02 14:30:33 [INFO] 3715 documents indexed from topic {remark https://remark42.com/demo13/}
remark42-dev  | 2023/04/02 14:30:33 [INFO] 3710 documents indexed from topic {remark https://remark42.com/demo4/}
remark42-dev  | 2023/04/02 14:30:33 [INFO] 3728 documents indexed from topic {remark https://remark42.com/demo7/}
remark42-dev  | 2023/04/02 14:30:33 [INFO] 3664 documents indexed from topic {remark https://remark42.com/demo5/}
remark42-dev  | 2023/04/02 14:30:33.062 [INFO]  {service/service.go:981 service.(*DataStore).IndexSites} finish building search index in 12.475010151s
remark42-dev  | 2023/04/02 14:30:33.062 [INFO]  {api/rest.go:116 api.(*Rest).Run} activate http rest server on :8080
remark42-dev  | 2023/04/02 14:30:33.062 [INFO]  {image/image.go:182 image.(*Service).Cleanup} start pictures cleanup, staging ttl=7m30s
remark42-dev  | 2023/04/02 14:30:33.064 [INFO]  {api/rest.go:483 api.addFileServer} run file server from ./web from the disk
```

</details>

paskal commented:
Can you please test on the cheapest droplet?

vdimir commented:
I tried on a small droplet (1gb ram) and faced memory issues, I will investigate.

vdimir commented:
@paskal 

Tested on latest commit on cheapest 512ram droplet 100k comments in 20 topics:

```
remark42-dev  | 2023/04/08 15:25:45.520 [INFO]  {service/service.go:974 service.(*DataStore).IndexSites} finish building search index in 1m56.786843857s
```

It seems it was an issue with too big batches passed to `bleve.Index` method (the whole topic). Now it's limited to 1024. 
Also, I made this method single-threaded because on multicore machines, it's fast enough even with one thread, but on slow, it may get stuck if it's run several threads.

paskal commented:
Are default concurrency settings good enough not to overuse the machine's memory?

vdimir commented:
> Are default concurrency settings good enough not to overuse the machine's memory?

I think so, peak usage was about 300-400mb.

I also tested on another workload with 1000 topics and ~100 comments each. It took more time - 4 min on 512mb ram single-core droplet.

```
...
remark42-dev  | 2023/04/10 14:16:03 [INFO] 95 documents indexed from topic {remark https://remark42.com/demo997/}
remark42-dev  | 2023/04/10 14:16:03 [INFO] 96 documents indexed from topic {remark https://remark42.com/demo998/}
remark42-dev  | 2023/04/10 14:16:03 [INFO] 94 documents indexed from topic {remark https://remark42.com/demo999/}
remark42-dev  | 2023/04/10 14:16:03.777 [INFO]  {service/service.go:974 service.(*DataStore).IndexSites} finish building search index in 4m3.937382438s
```
